src/Search/Services/SearchableService.php

```diff
@@ -0,0 +1,166 @@
+     */
+    public function isIndexable(DataObject $obj): bool
+    {
+        return $this->isSearchable($obj, true);
+    }
+
+    /**
+     * Used when retrieving search results
+     *
+     * This is considered the secondary layer of protection
+     *
+     * It's important to still have this layer in conjuction with the index layer as non-searchable results may be
+     * in the search index because:
+     * a) they were added to the index pre-fulltextsearch 3.7 and a reindex to purge old records was never run, OR
+     * b) the DataObject has a non-deterministic canView() check such as `return $date <= $dateOfIndex;`
+     *
+     * @param DataObject $obj
+     * @return bool
+     */
+    public function isViewable(DataObject $obj): bool
+    {
+        return $this->isSearchable($obj, false);
+    }
+
+    /**
+     * Checks and caches whether the given DataObject can be indexed. This is determined by two factors:
+     * - Whether the ShowInSearch property / getShowInSearch() method evaluates to true
+     * - Whether the canView method evaluates to true against an anonymous user (optional, can be disabled)
+     *
+     * @param DataObject $obj
+     * @param bool $indexing
+     * @return bool
+     */
+    private function isSearchable(DataObject $obj, bool $indexing): bool
+    {
+        // check if is a valid DataObject that has been persisted to the database
+        if (is_null($obj) || !$obj->ID) {
+            return false;
+        }
+
+        $key = $this->getCacheKey($obj, $indexing);
+        if (isset($this->cache[$key])) {
+            return $this->cache[$key];
+        }
+
+        $value = true;
+
+        // ShowInSearch check
+        // This will also call $obj->getShowInSearch() if it exists
+        if (isset($obj->ShowInSearch) && !$obj->ShowInSearch) {
+            $value = false;
+        }
+
+        // canView() checker
+        if ($value) {
+            $objClass = $obj->getClassName();
+            if ($indexing) {
+                // Anonymous member canView() for indexing
+                if (!$this->classSkipsCanViewCheck($objClass)) {
+                    $value = Member::actAs(null, function () use ($obj) {
+                        return $obj->canView();
+                    });
+                }
+            } else {
+                // Current member canView() check for retrieving search results
+                $value = $obj->canView();
+            }
+        }
+        $this->extend('updateIsSearchable', $obj, $indexing, $value);
+        $this->cache[$key] = $value;
+        return $value;
+    }
+
+    /**
+     * Generates a unique key to cache each DataObject with. Can be extended via updateCacheKey.
+     *
+     * @param DataObject $obj
+     * @param bool $indexing
+     * @return string
+     */
+    private function getCacheKey(DataObject $obj, bool $indexing): string
+    {
+        $key = implode('_', [$obj->ClassName, $obj->ID, $indexing ? 'indexing' : 'viewing']);
```

Comment:
Should utilise https://github.com/silverstripe/silverstripe-framework/pull/9400/files which is close to being merged

Reply:
Bear in mind that you'll also need to bump the minimum core requirement for this module to 4.6 in order to adopt that API.

> you'll never do both in the same request

Never say never 😉

Reply:
Fine I'll keep it :-) Will simply update to use `$obj->getUniqueKey()`